lib/teleterm/daemon/daemon_headless_watcher.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package daemon
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+// StartHeadlessHandlers starts a headless watcher for the given cluster URI.
+func (s *Service) StartHeadlessWatcher(uri string) error {
+	s.headlessWatcherClosersMu.Lock()
+	defer s.headlessWatcherClosersMu.Unlock()
+
+	cluster, err := s.ResolveCluster(uri)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := s.startHeadlessWatcher(cluster); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(err)
+}
+
+// StartHeadlessWatchers starts headless watchers for all connected clusters.
+func (s *Service) StartHeadlessWatchers() error {
+	s.headlessWatcherClosersMu.Lock()
+	defer s.headlessWatcherClosersMu.Unlock()
+
+	clusters, err := s.cfg.Storage.ReadAll()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for _, c := range clusters {
+		if c.Connected() {
+			if err := s.startHeadlessWatcher(c); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// StopHeadlessWatcher stops the headless watcher for the given cluster URI.
+func (s *Service) StopHeadlessWatcher(uri string) error {
+	s.headlessWatcherClosersMu.Lock()
+	defer s.headlessWatcherClosersMu.Unlock()
+
+	closer, ok := s.headlessWatcherClosers[uri]
+	if !ok {
+		return trace.NotFound("no headless watcher for cluster %v", uri)
+	}
+
+	closer()
+	delete(s.headlessWatcherClosers, uri)
+	return nil
+}
+
+// StopHeadlessWatchers stops all headless watchers.
+func (s *Service) StopHeadlessWatchers() {
+	s.headlessWatcherClosersMu.Lock()
+	defer s.headlessWatcherClosersMu.Unlock()
+
+	for uri, closer := range s.headlessWatcherClosers {
+		closer()
+		delete(s.headlessWatcherClosers, uri)
+	}
+}
+
+// startHeadlessWatcher starts a background process to watch for pending headless
+// authentications for this cluster.
+func (s *Service) startHeadlessWatcher(cluster *clusters.Cluster) error {
+	if _, ok := s.headlessWatcherClosers[cluster.URI.String()]; ok {
```

Comment:
I think it won't work correctly in the following scenario:
1. Log in to a cluster.
2. Wait for a cert to expire.
3. Try to log in again.

Step 3 will fail with an error saying that the watcher already exists, won't it? I guess we should just close the old watcher before starting the new one.

Reply:
Good point. I think catching `AlreadyExists` in `StartHeadlessWatcher` and simply logging the fact should do the job. Its plural equivalent, `StartHeadlessWatchers`, should definitely error if a watcher already exists.

Reply:
I'll update it so that when the watcher goroutine loop ends, it is delete from the map. This way, if the certs expire, the watcher should clean itself up.

Reply:
You're right...I forget that a client can still function with expired certs, just can't redial. Updated to have `StartHeadlessWatcher` stop and replace the previous watcher if there is one.

Reply:
The watcher will still work and send the headless authn to the Electron App. If the user hasn't logged in however, the Electron app will fail to approve/deny the headless authn, trigger relogin, and retry. At this point, the headless authn can be handled and the old headless watcher will be replaced. So in a roundabout way, the Headless watcher system should handle relogin surprisingly well.

Reply:
Right, until the client is closed or receives a connection error.